src/flp/types_utilities.rs

```diff
@@ -0,0 +1,142 @@
+
+//! Some functions shared by several [`Type`](crate::flp::Type) implementations.
+
+use crate::field::FieldElement;
+use crate::flp::{FlpError, Gadget, Type};
+use std::convert::TryFrom;
+
+/// Interpret `bits` as `F::Integer` if it's representable in that type and smaller than the field modulus.
+pub(crate) fn valid_int_from_usize<F: FieldElement>(bits: usize) -> Result<F::Integer, FlpError> {
+    let bits_int = F::Integer::try_from(bits).map_err(|err| {
+        FlpError::Encode(format!(
+            "bit length ({}) cannot be represented as a field element: {:?}",
+            bits, err,
+        ))
+    })?;
+
+    if F::modulus() >> bits_int == F::Integer::from(F::zero()) {
+        return Err(FlpError::Encode(format!(
+            "bit length ({}) exceeds field modulus",
+            bits,
+        )));
+    }
+    Ok(bits_int)
+}
+
+/// Encode `input` as `bits`-bit vector of elements of `T::Field` if it's small enough
+/// to be represented with that many bits.
+///
+/// # Arguments
+///
+/// * `input` - The field element to encode
+/// * `bits` - The number of bits to use for the encoding
+/// * `max_input` - The maximal number representable with `bits` bits. Equal to `bits^2-1`
+pub(crate) fn encode_into_bitvector_representation<F: FieldElement>(
+    input: &F::Integer,
+    bits: usize,
+    max_input: F::Integer,
+) -> Result<Vec<F>, FlpError> {
+    if *input > max_input {
+        return Err(FlpError::Encode(
+            "value of input exceeds bit length".to_string(),
+        ));
+    }
+
+    let one = F::Integer::from(F::one());
+    let mut encoded: Vec<F> = Vec::with_capacity(bits);
+    for l in 0..bits {
+        let l = F::Integer::try_from(l).unwrap();
+        let w = F::from((*input >> l) & one);
+        encoded.push(w);
+    }
+    Ok(encoded)
+}
+
+/// Decode the bitvector-represented value `input` into a simple representation as a single field element.
+pub(crate) fn decode_from_bitvector_representation<F: FieldElement>(input: &[F]) -> F {
+    let mut decoded = F::zero();
+    for (l, bit) in input.iter().enumerate() {
+        let w = F::from(F::Integer::try_from(1 << l).unwrap());
+        decoded += w * *bit;
+    }
+    decoded
+}
+
+/// Compute a random linear combination of the result of calls of `g` on each element of `input`.
+///
+/// # Arguments
+///
+/// * `g` - The gadget to be applied elementwise
+/// * `input` - The vector on whose elements to apply `g`
+/// * `rnd` - The randomness used for the linear combination
+pub(crate) fn call_gadget_on_vec_entries<F: FieldElement>(
+    g: &mut Box<dyn Gadget<F>>,
+    input: &[F],
+    rnd: F,
+) -> Result<F, FlpError> {
+    let mut range_check = F::zero();
+    let mut r = rnd;
+    for chunk in input.chunks(1) {
+        range_check += r * g.call(chunk)?;
+        r *= rnd;
+    }
+    Ok(range_check)
+}
+
+/// Check whether `input` and `joint_rand` have the length expected by `typ`.
+pub(crate) fn valid_call_check<T: Type>(
```

Comment:
Similar to my suggestion of adding default implementations to `trait FieldElement`, I think this function and `truncate_call_check` could be refactored into methods on `trait Type` in `flp.rs`. In this case, I argue for _methods_ and not associated functions since these two new functions already take `typ: &T` and so that could become `&self`. e.g:

```
    /// Check whether `input` and `joint_rand` have the length expected by `self`.
    fn valid_call_check(
        &self,
        input: &[Self::Field],
        joint_rand: &[Self::Field],
    ) -> Result<(), FlpError> {
        if input.len() != self.input_len() {
            return Err(FlpError::Valid(format!(
                "unexpected input length: got {}; want {}",
                input.len(),
                self.input_len(),
            )));
        }

        if joint_rand.len() != self.joint_rand_len() {
            return Err(FlpError::Valid(format!(
                "unexpected joint randomness length: got {}; want {}",
                joint_rand.len(),
                self.joint_rand_len()
            )));
        }

        Ok(())
    }
```